internal/keepers/ocr.go

```diff
@@ -100,38 +100,67 @@ func (k *keepers) Report(ctx context.Context, rt types.ReportTimestamp, _ types.
 		return false, nil, fmt.Errorf("%w: failed to sort/dedupe attributed observations: %s", err, lCtx)
 	}
 
-	// TODO: Generate multiple reports
-	// select, verify, and build report
-	toPerform := make([]ktypes.UpkeepResult, 0, 1)
-	for _, key := range keys {
-		upkeeps, err := k.service.CheckUpkeep(ctx, key)
-		if err != nil {
-			return false, nil, fmt.Errorf("%w: failed to check upkeep from attributed observation: %s", err, lCtx)
-		}
+	// No keys found for the given keys
+	if len(keys) == 0 {
+		k.logger.Printf("OCR report completed successfully with no eligible keys: %s", lCtx)
+		return false, nil, nil
+	}
 
-		// No upkeeps found for the given key
-		if len(upkeeps) == 0 {
-			continue
-		}
+	// Check all upkeeps from the given observation
+	checkedUpkeeps, err := k.service.CheckUpkeep(ctx, keys...)
+	if err != nil {
+		return false, nil, fmt.Errorf("%w: failed to check upkeeps from attributed observation: %s", err, lCtx)
+	}
 
-		if len(upkeeps) > 1 {
-			return false, nil, fmt.Errorf("unexpected number of upkeeps returned for %s key, expected 1 but given %d", key, len(upkeeps))
-		}
+	// No upkeeps found for the given keys
+	if len(checkedUpkeeps) == 0 {
+		k.logger.Printf("OCR report completed successfully with no successfully checked upkeeps: %s", lCtx)
+		return false, nil, nil
+	}
 
-		if upkeeps[0].State == ktypes.Eligible {
-			// only build a report from a single upkeep for now
-			k.logger.Printf("reporting %s to be performed: %s", upkeeps[0].Key, lCtx.Short())
-			toPerform = append(toPerform, upkeeps[0])
-			break
+	if len(checkedUpkeeps) > len(keys) {
+		return false, nil, fmt.Errorf("unexpected number of upkeeps returned for %s key, expected max %d but given %d", key, len(keys), len(checkedUpkeeps))
+	}
+
+	// Collect eligible upkeeps
+	keyToUpkeep := make(map[string]ktypes.UpkeepResult)
+	eligibleUpkeeps := make([]ktypes.UpkeepKey, 0, len(checkedUpkeeps))
+	for _, checkedUpkeep := range checkedUpkeeps {
+		if checkedUpkeep.State == ktypes.Eligible {
+			eligibleUpkeeps = append(eligibleUpkeeps, checkedUpkeep.Key)
+			keyToUpkeep[string(checkedUpkeep.Key)] = checkedUpkeep
 		}
 	}
 
 	// if nothing to report, return false with no error
-	if len(toPerform) == 0 {
-		k.logger.Printf("OCR report completed successfully with no upkeeps added to the report: %s", lCtx)
+	if len(eligibleUpkeeps) == 0 {
+		k.logger.Printf("OCR report completed successfully with no eligible upkeeps: %s", lCtx)
 		return false, nil, nil
 	}
 
+	// Get upkeep gas limit
+	upkeepsInfo, err := k.service.GetUpkeep(ctx, eligibleUpkeeps...)
+	if err != nil {
+		return false, nil, fmt.Errorf("%w: failed to get upkeeps info: %s", err, lCtx)
+	}
```

Comment:
Will it be too complex if we try to batch this call within checkUpkeep itself? Would save us sequential RPC calls.

Reply:
yeah. Maybe it's ok for now and in next version we return the gas limit within simulatePerformUpkeep. wdyt @EasterTheBunny ?

Reply:
I think we should think through this one a bit more. My big concern is that checking upkeeps and getting upkeeps (if they are both returning different data) are occurring in sequence instead of parallel. Let's go back to what data we need exactly, why it's different from checking upkeeps, and see if we can put that data call somewhere else.

Reply:
do we have any plans to make this service stateful? gasLimit seems like a prime contender for a piece of data that could live in a db. also agreed that returning from simulatePerform sounds useful.

Reply:
@RyanRHall yes and that's something I'm working on a bit in the chainlink repo. Roman and I are going to discuss it today.